tools/roofline.py

```diff
@@ -0,0 +1,126 @@
+
+class Roofline:
+    def __init__(self, streaming_limit, flop_limit):
+        """The generation of a roofline performance model, for given code.
+
+        :arg self: self
+        :arg streaming_limit: Memory Streaming Bandwidth (GB/s)
+        :arg flop_limit: CPU's Maximum Memory Bandwidth (GB/s)
+        """
+        self.data = defaultdict(partial(defaultdict, partial(defaultdict, float)))
+        self.streaming_limit = streaming_limit
+        self.flop_limit = flop_limit
+    
+    def start_collection(self, region_name=None):
+        """The start point of data collection for the Roofline model."""
+        start = PETSc.Log.getPerfInfoAllStages()['Main Stage']
+        data = self.data[region_name]
+        for event, info in start.items():
+            event_data = data[event]
+            for n in ('flops', 'bytes', 'time', 'count'):
+                event_data[n] -= info[n]
+
+    def stop_collection(self, region_name=None):
+        """The end point of data collection for the Roofline model."""
+        stop = PETSc.Log.getPerfInfoAllStages()['Main Stage']
+        data = self.data[region_name]
+        for event, info in stop.items():
+            event_data = data[event]
+            for n in ('flops', 'bytes', 'time', 'count'):
+                event_data[n] += info[n]
+
+    @contextmanager
+    def collecting(self, region_name=None):
+        """Automated inclusion of stop_collection at the end of a script if not called."""
+        self.start_collection(region_name)
+        try: 
+            yield
+        finally:
+            self.stop_collection(region_name)
+
+    def roofline(self, region_name=None, event_name=None, axes=None):
+        """The generation of a roofline plot.
+
+        :arg self: Self
+        :arg data_type: Choice between 'flops', 'bytes', and 'time'
+        :arg axes: Existing axes to add roofline plot to
+        :arg data: Load previously saved data stored as a pickle file
+        :returns: Roofline plot axes
+        """
+
+        if axes is None:
+            figure = plt.figure()
+            axes = figure.add_subplot(111)
+
+        if event_name is not None:
+            data = self.data[region_name][event_name]
+        else: 
+            data = defaultdict(float)
+            for event in self.data[region_name].values():
+                data['flops'] += event['flops']
+                data['bytes'] += event['bytes'] 
+                data['time'] += event['time']
+
+        intensity = data['flops']/data['bytes']
+        flop_rate = (data['flops']/data['time']) * 1e-9
+        
+        x_range = [-6, 10] 
+        x = numpy.logspace(x_range[0], x_range[1], base=2, num=100)
+        y = []
+        for points in x: 
+            # The minimum of the memory streaming bandwidth and compute limit
+            y.append(min(points * self.streaming_limit, self.flop_limit))
+        
+        mem_lim = self.flop_limit/self.streaming_limit
+        x_mem, y1_mem, y2_mem = [], [], []
```

Comment:
What are you doing in this bit? Please can you add some documentation/more informative variable names.